content/en/blog/_posts/2022-12-01-scalable-job-tracking-ga/index.md

```diff
@@ -0,0 +1,132 @@
+and [failure handling](/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)
+policies, requiring the end state of a finished Pod to determine whether to
+create a replacement Pod or mark the Job as completed or failed. As a result,
+the Job controller depended on Pods, even terminated ones, to remain in the API
+in order to keep track of the status.
+
+This dependency made the tracking of Job status unreliable, because Pods can be
+deleted from the API for a number of reasons, including:
+- The garbage collector removing orphan Pods when a Node goes down.
+- The garbage collector removing terminated Pods when they reach a treshhold.
+- The Kubernetes scheduler preempting a Pod to accomodate higher priority Pods.
+- The taint manager evicting a Pod that doesn't tolerate a `NoExecute` taint.
+- External controllers, not included as part of Kubernetes, or humans deleting
+  Pods.
+
+## How did we fix the Job controller?
+
+When a controller needs to take an action on deleted objects, it should add a
+[finalizer](/docs/concepts/overview/working-with-objects/finalizers/) to the
+objects that it manages.
+A finalizer prevent the objects from being deleted from the API until the
+finalizers are removed. Once the controller is done doing cleanup and accounting
+for the deleted object, it can remove the finalizer from the object and the
+control plane removes the object from the API.
+
+This is what the new Job controller is doing: adding a finalizer during Pod
+creation, and removing it after the Pod terminates and is accounted for in the
+Job status. However, it wasn't that simple.
+
+The main challenge is that there are at least two objects involved: the Pod
+and the Job. While the finalizer lives in the Pod object, the accounting lives
+in the Job object. There is no mechanism to atomically remove the finalizer in
+the Pod and update the counters in the Job status. And there could be more than
+one terminated Pod at a given time.
+
+To solve this problem, we implemented a three staged approach, each translating
+to an API call.
+1. Add the terminated Pod(s) UID into temporary lists in the Job status
+   ([uncountedTerminatedPods](/docs/reference/kubernetes-api/workload-resources/job-v1/#JobStatus)).
+2. Remove the finalizer from the Pods(s).
+3. Atomically remove UIDs from the temporary lists and increment the `succeeded`
+   and `failed` counters.
+
+Additional complications come from the fact that the Job controller might
+receive the results of the API changes in steps 1 and 2 out of order. We solved
+this by adding an in-memory cache for removed finalizers.
+
+Still, we faced some issues during the beta stage that prompted us to disable
+the feature in 1.23 and 1.24 ([#108645](https://github.com/kubernetes/kubernetes/issues/108645),
+[#109485](https://github.com/kubernetes/kubernetes/issues/109485),
+[#111646](https://github.com/kubernetes/kubernetes/pull/111646)).
```

Comment:
I'm not sure we need to include the off-again/on-again bits in the blog post.

Reply:
I thought it could be valuable information about how complex the feature was so that developers think twice before trying to manage pods themselves instead of using the Job API.

Reply:
Maybe instead of pointing them to issue write down a sentence or two about the issues you've faced and link to issue for more details.